File: backend/app/services/leak_analysis.py
import logging
from typing import List, Dict, Any
from motor.motor_asyncio import AsyncIOMotorDatabase
from bson import ObjectId

logger = logging.getLogger("app.services.leak_analysis")

def get_severity_badge(pct: float) -> str:
    if pct >= 30.0:
        return "High"
    elif pct >= 10.0:
        return "Medium"
    return "Low"

async def get_ranked_leak_points(db: AsyncIOMotorDatabase, assessment_id: str) -> List[Dict[str, Any]]:
    """
    Ranks emission leak points for an assessment. Reads calculated computed_co2e values, writes nothing back.
    """
    cursor = db.process_inputs.find({"assessment_id": assessment_id})
    inputs = await cursor.to_list(length=500)

    if not inputs:
        return []

    # Aggregate by subtype / input name
    aggregated = {}
    total_co2e = 0.0

    for item in inputs:
        co2e = float(item.get("computed_co2e", 0.0))
        subtype = item.get("subtype", "unknown")
        category = item.get("category", "general")

        total_co2e += co2e
        key = (category, subtype)
        if key not in aggregated:
            aggregated[key] = {
                "category": category,
                "subtype": subtype,
                "name": subtype.replace("_", " ").title(),
                "computed_co2e": 0.0
            }
        aggregated[key]["computed_co2e"] += co2e

    # Build ranked list
    items_list = list(aggregated.values())
    for item in items_list:
        co2e = item["computed_co2e"]
        pct = round((co2e / total_co2e * 100.0), 2) if total_co2e > 0 else 0.0
        item["contribution_pct"] = pct
        item["computed_co2e"] = round(co2e, 4)

    # Sort descending
    items_list.sort(key=lambda x: x["contribution_pct"], reverse=True)

    ranked_leak_points = []
    for rank, item in enumerate(items_list, start=1):
        pct = item["contribution_pct"]
        ranked_leak_points.append({
            "rank": rank,
            "name": item["name"],
            "category": item["category"],
            "subtype": item["subtype"],
            "computed_co2e": item["computed_co2e"],
            "contribution_pct": pct,
            "severity_badge": get_severity_badge(pct)
        })

    return ranked_leak_points

File: backend/src/services/leakAnalysis.js
const ProcessInput = require('../models/ProcessInput');
const Assessment = require('../models/Assessment');

/**
 * Determine severity badge based on percentage contribution
 */
function getSeverity(pctContribution) {
  if (pctContribution >= 0.30) return 'high';
  if (pctContribution >= 0.10) return 'medium';
  return 'low';
}

/**
 * Rank emission leak points for an assessment
 * Aggregates by (category, subtype), calculates contribution, assigns severity
 */
async function getLeakPoints(assessmentId) {
  const assessment = await Assessment.findById(assessmentId);
  if (!assessment) throw { code: 'NOT_FOUND', message: 'Assessment not found', status: 404 };
  if (assessment.status !== 'complete') {
    throw { code: 'ASSESSMENT_NOT_CALCULATED', message: 'Assessment has not been calculated yet', status: 400 };
  }

  const inputs = await ProcessInput.find({
    assessment_id: assessmentId,
    computed_co2e: { $ne: null },
    unsupported: false,
  });

  if (!inputs.length) {
    return { leak_points: [] };
  }

  // Aggregate by (category, subtype)
  const aggregated = {};
  for (const input of inputs) {
    const key = `${input.category}__${input.subtype}`;
    if (!aggregated[key]) {
      aggregated[key] = {
        category: input.category,
        subtype: input.subtype,
        co2e: 0,
      };
    }
    aggregated[key].co2e += input.computed_co2e || 0;
  }

  const totalCo2e = assessment.total_co2e || 1;

  // Build ranked list
  const leakPoints = Object.values(aggregated)
    .map((lp) => {
      const pct = totalCo2e > 0 ? lp.co2e / totalCo2e : 0;
      return {
        leak_point_ref: lp.subtype,
        category: lp.category,
        subtype: lp.subtype,
        co2e: parseFloat(lp.co2e.toFixed(4)),
        pct_contribution: parseFloat(pct.toFixed(4)),
        severity: getSeverity(pct),
      };
    })
    .sort((a, b) => b.co2e - a.co2e)
    .map((lp, idx) => ({ rank: idx + 1, ...lp }));

  return { leak_points: leakPoints };
}

module.exports = { getLeakPoints, getSeverity };
